### PR TITLE
Feature/TAN-5 CRUD for organizations

### PR DESCRIPTION
**Pull request by misstonbon**
Sets up CRUD actions for organizations, and organizations reducer.

### CODE REVIEW

**Comment by kcoleman731** [comment]
@misstonbon would love for you to add detailed descriptions for all PRs. For instance...

This pull requests adds support for CRUD organization actions in addition to an organization reducer.

File: src/actions/organization.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import { ActionCreatorsMapObject, Dispatch, Action } from 'redux'
+import { IOrganization } from '../reducers/organization'
+
+// API Helpers
+import { POST } from '../helpers/api'
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: consider using `import * as api from '../helpers/api'`. Always bias towards writing less code.

File: src/actions/organization.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import { PUT } from '../helpers/api'
+import { GET } from '../helpers/api'
+import { DELETE } from '../helpers/api'
+
+export const FETCHED_ORGANIZATIONS = 'FETCH_ORGANIZATIONS'
+export const CREATE_ORGANIZATION = 'CREATE_ORGANIZATION'
+export const CREATED_ORGANIZATION = 'CREATED_ORGANIZATION'
+
+export const DELETE_ORGANIZATION = 'DELETE_ORGANIZATION'
+export const DELETED_ORGANIZATION = 'DELETE_ORGANIZATION'
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: Check you `DELETED_ORGANIZATION` string.

File: src/actions/organization.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+// -----------------------------------------------------------------------------
+// Get all Organizations
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: `GET Organizations`

File: src/actions/organization.tsx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * Gets all organizations via a call to the `/organizations` API.
+ */
+const fetchOrganizations = () => (dispatch: Dispatch<Action>) => {
+	return GET('/organizations', {}).then((organizations: IOrganization[]) => {
+		dispatch(fetchedOrganizations(organizations))
+	}).catch((err) => Promise.reject(err))
+}
+/**
+ * Builds a `IFetchedOrganizationsAction ` upon successful fetch.
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: Builds `an`

File: src/actions/organization.tsx [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+}
+/**
+ * IFetchedOrganizationsAction is dispatched after organizations are fethed.
+ */
+export interface IFetchedOrganizationsAction extends Action {
+	type: string,
+	organizations: IOrganization[],
+	receivedAt: number
+}
+// -----------------------------------------------------------------------------
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: add new line.

File: src/actions/organization.tsx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+// Create or update organization depending on button type
+// -----------------------------------------------------------------------------
+/**
+ * Creates a new organization via a call to the `/organizations` API.
+ * @param name The name supplied in the organization form.
+ * @param id Auto-assigned id number.
+ */
+const createOrganization = ( id: number, name: string ) => (dispatch: Dispatch<ICreatedOrganizationAction>) => {
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: We want to have separate functions for creating and updating a functions. So we want to have `createOrganizations` and `updateOrganization`.

File: src/actions/organization.tsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	organizations: IOrganization[],
+	receivedAt: number
+}
+// -----------------------------------------------------------------------------
+// Create or update organization depending on button type
+// -----------------------------------------------------------------------------
+/**
+ * Creates a new organization via a call to the `/organizations` API.
+ * @param name The name supplied in the organization form.
+ * @param id Auto-assigned id number.
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: Your parameters are out of order here.

File: src/actions/organization.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+ */
+const createOrganization = ( id: number, name: string ) => (dispatch: Dispatch<ICreatedOrganizationAction>) => { 
+	// since this handles edit and create, could be named better.
+	const organizationPayload = { id, name }
+	const func = this.state.buttonType === 'edit' ? PUT : POST
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: different functions means we won't need this line.

File: src/actions/organization.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	return GET('/organizations', {}).then((organizations: IOrganization[]) => {
+		dispatch(fetchedOrganizations(organizations))
+	}).catch((err) => Promise.reject(err))
+}
+/**
+ * Builds a `IFetchedOrganizationsAction ` upon successful fetch.
+ * @param organizations The response from the organizations API call.
+ * @return The `FetchedOrganizationsAction` instance.
+ */
+const fetchedOrganizations = (organizations: IOrganization[]): Action => {
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: check your return type.

File: src/actions/organization.tsx [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+}
+/**
+ * ICreatedOrganizationAction is dispatched after organization has been created.
+ */
+export interface ICreatedOrganizationAction extends Action {
+	type: string,
+	organization: IOrganization,
+	receivedAt: number
+}
+// -----------------------------------------------------------------------------
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: new line.

File: src/actions/organization.tsx [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+ */
+const createdOrganization = (organization: IOrganization): ICreatedOrganizationAction => {
+	const action: ICreatedOrganizationAction = {
+		organization: organization,
+		receivedAt: Date.now(),
+		type: CREATED_ORGANIZATION,
+	}
+	return action
+}
+/**
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: new line

File: src/actions/organization.tsx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+const createOrganization = ( id: number, name: string ) => (dispatch: Dispatch<ICreatedOrganizationAction>) => { 
+	// since this handles edit and create, could be named better.
+	const organizationPayload = { id, name }
+	const func = this.state.buttonType === 'edit' ? PUT : POST
+	// assuming we have edit buttonType
+	return func('/organizations', organizationPayload).then((organization: IOrganization) => {
+		dispatch(createdOrganization(organization))
+	}).catch((err) => Promise.reject(err))
+}
+/**
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: new line

File: src/actions/organization.tsx [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+ */
+export interface ICreatedOrganizationAction extends Action {
+	type: string,
+	organization: IOrganization,
+	receivedAt: number
+}
+// -----------------------------------------------------------------------------
+// Delete organization
+// -----------------------------------------------------------------------------
+const deleteOrganization( organization: IOrganization ) => (dispatch: Dispatch<Action>) => {
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: new line and docs!

File: src/actions/organization.tsx [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+	return DELETE('/organization/', organization.id).then((organizations: IOrganization[]) => {
+		dispatch(deletedOrganization(organizations))
+	}).catch((err) => Promise.reject(err))
+}
+
+const deletedOrganization = (organizations: IOrganization[]): Action => {
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: docs

File: src/actions/organization.tsx [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+	const action: IDeletedOrganizationAction = { 
+		// I'm confused. I want this to serve as "successful delete object" that updates organizations state reflectting deletion of organization
+		organizations: organizations,
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: It all depends on what object comes back from the delete call. Ideally nothing should come back. Then you can removed the delete object from your state in the reducer.

File: src/actions/organization.tsx [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+	organizations: IOrganization[],
+	receivedAt: number
+}
+/**
+ * Defines the interface for our OrganizationAction object.
+ */
+export interface OrganizationDispatch extends ActionCreatorsMapObject {
+	createOrganization(id: number, name: string ): (dispatch: Dispatch<ICreatedOrganizationAction>) => Promise<void>
+	createdOrganization(organization: IOrganization): ICreatedOrganizationAction
+	deleteOrganization(organizations: IOrganization[]): IDeletedOrganizationAction
```
Comment (kcoleman731) [inline src/actions/organization.tsx]: you are missing deletedOrganizaton

File: src/helpers/api.tsx [hunk 17]
```diff
@@ -59,7 +59,7 @@ export function DELETE(path: string, params: any) {
 function performFetch(url: string, options: RequestInit) {
 	console.log('URL', url)
 	console.log('Options', options)
-	return fetch(url, options).then((response: Response) => {
+	return fetch(url, options).then((response: any) => {
```
Comment (kcoleman731) [inline src/helpers/api.tsx]: Avoid the use of Anys at all cost!

File: src/reducers/organization.tsx [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+function organization(state = defaultState, action: AnyAction): IOrganizationState {
+	let typedAction
+	switch (action.type) {
+		case FETCHED_ORGANIZATIONS:
+			return {
+				...state,
+				organizations: action.organizations
+			}
+		case CREATED_ORGANIZATION:
+			typedAction = action as ICreatedOrganizationAction
```
Comment (kcoleman731) [inline src/reducers/organization.tsx]: You aren't ever using the typedAction. You should however.

Comment (misstonbon) [reply src/reducers/organization.tsx]: added

File: __tests__/actions/organization.test.ts [hunk 19]
```diff
@@ -0,0 +1,176 @@
+import { expect } from 'chai'
+import configureMockStore, { MockStoreEnhanced, MockStoreCreator } from 'redux-mock-store'
+import thunk, { ThunkDispatch } from 'redux-thunk'
+import { Action, Middleware, AnyAction } from 'redux'
+
+import {
```
Comment (kcoleman731) [inline __tests__/actions/organization.test.ts]: whenever you have a bunch of imports like this from the same file, better to do `import * as orgActions from 'from '../../src/actions/organization`

File: __tests__/reducers/organization.test.ts [hunk 20]
```diff
@@ -0,0 +1,88 @@
+import { expect } from 'chai'
+import { Action, AnyAction } from 'redux'
+
+import OrganizationReducer, { IOrganizationState } from '../../src/reducers/organization'
+import {
+		ICreatedOrganizationAction,
```
Comment (kcoleman731) [inline __tests__/reducers/organization.test.ts]: same comment here.

File: src/reducers/organization.tsx [hunk 21]
```diff
@@ -0,0 +1,61 @@
+import { AnyAction } from 'redux'
+import {
```
Comment (kcoleman731) [inline src/reducers/organization.tsx]: same here.

**Pull request merged**